content/influxdb/cloud-dedicated/query-data/execute-queries/optimize-queries.md

```diff
@@ -29,7 +31,25 @@ Use the following tools to help you identify performance bottlenecks and trouble
   - [Inspect Flight response headers](#inspect-flight-response-headers)
 - [Retrieve query information](#retrieve-query-information)
 
-<!-- /TOC -->
+
+## Strategies for improving query performance
+
+A query may be slow due to the following reasons:
+
+- It queries a large time-range of data.
+- It includes intensive operations, such as `ORDER BY` on large amounts of data.
+- The query plan isn't optimal--for example, applying the same sort (`ORDER BY`) to already sorted data.
```

Comment:
`DeduplicateExec` indicates expensive operation due to overlapped files. A few is OK, > 10 files, generally not good.
Total number of file groups under `Union` is additive (split and then unioned.
`SortPreserveMergeExec` indicates the file is already sorted (by tags and time) and doesn't need re-sorting.
`SortExec` indicates the file is sorted again;`SortExec` within a `DeduplicateExec` is a sign of a more expensive operation.
# of file groups can be configured to determine when files are processed in parallel vs. sequentially.

Reply:
@NGA-TRAN

Reply:
A few samples

```SQL
ProjectionExec: expr=[cluster_id@1 as cluster_id, device@2 as device, free@3 as free, fstype@4 as fstype, host@5 as host, inodes_free@6 as inodes_free, inodes_total@7 as inodes_total, inodes_used@8 as inodes_used, mode@9 as mode, path@10 as path, time@11 as time, total@12 as total, used@13 as used, used_percent@14 as used_percent]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |         DeduplicateExec: [cluster_id@1 ASC,host@5 ASC,time@11 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |           SortPreservingMergeExec: [cluster_id@1 ASC,host@5 ASC,time@11 ASC,__chunk_order@0 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |             SortExec: expr=[cluster_id@1 ASC,host@5 ASC,time@11 ASC,__chunk_order@0 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |               CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |                 FilterExec: time@11 >= 1687885391398358371 AND (time@11 < -9223372036854775808 OR time@11 > 1656349451399264154)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |                   ParquetExec: file_groups={8 groups: [[5/226/34124/3eb6e785-ef65-432d-9069-00bb55d24607.parquet, 5/226/34124/39c58ab8-0e8a-4069-a800-b08b0244fbe1.parquet, 5/226/34124/c366b94a-4b81-4d23-8696-a415a5dd55e4.parquet, 5/226/34124/f8931b61-e126-44a7-9d87-c97372df83eb.parquet, 5/226/34124/4ff24452-032c-418d-ad3e-3bf736d568f0.parquet, ...], [5/226/34124/5af29749-b6e4-4b53-b14a-b6fbafb1c03a.parquet, 5/226/34124/ff76afb5-dc4f-4b29-8de6-abfda799c74b.parquet, 5/226/34124/6609aa70-f028-42e8-b3b1-eac3a9bb5fb0.parquet, 5/226/34124/7ae63110-129d-4e2a-b651-ac796a8ed3fd.parquet, 5/226/34124/71eed42b-2673-4072-9d6e-159c637f7ce5.parquet, ...], [5/226/34124/3c7c00b8-6f7e-4038-972f-495faa3c9058.parquet, 5/226/34124/8cae1c49-8a52-44c5-899c-fd7f90fb5e98.parquet, 5/226/34124/2ab1b44a-6463-4b9c-a8bc-c08480845185.parquet, 5/226/34124/fc2ed189-4f76-40b2-8905-5e064a8d3b87.parquet, 5/226/34124/2aef3fbc-6666-4737-90ee-2ab08f4d0fc4.parquet, ...], [5/226/34124/5ef7958b-5ec2-4625-ad90-99ab22ca15fd.parquet, 5/226/34124/cc049840-2681-462c-b2d6-bb0cb27af07b.parquet, 5/226/34124/9964b17a-e714-4261-a627-96ad76b51731.parquet, 5/226/34124/c7c30233-30be-4be3-a4b1-8d43fa316645.parquet, 5/226/34124/aa3a8e5e-cbc1-4a73-a11f-9077cd20fef6.parquet, ...], [5/226/34124/7a39ee5d-0bba-4772-bbd0-8370ebb8cc89.parquet, 5/226/34124/4a09a54e-6347-47da-9c5b-d046477771a2.parquet, 5/226/34124/a039afff-cc4b-4100-8c82-ecce5bd71325.parquet, 5/226/34124/259c3d37-b9ee-4000-b6be-fee26f92ae2f.parquet, 5/226/34124/cfb500f5-bbdf-46c7-8d0c-83e599162433.parquet, ...], ...]}, projection=[__chunk_order, cluster_id, device, free, fstype, host, inodes_free, inodes_total, inodes_used, mode, path, time, total, used, used_percent], predicate=time@10 >= 1687885391398358371 AND (time@10 < -9223372036854775808 OR time@10 > 1656349451399264154), pruning_predicate=time_max@0 >= 1687885391398358371 AND (time_min@1 < -9223372036854775808 OR time_max@0 > 1656349451399264154) |
|               |       ProjectionExec: expr=[cluster_id@1 as cluster_id, device@2 as device, free@3 as free, fstype@4 as fstype, host@5 as host, inodes_free@6 as inodes_free, inodes_total@7 as inodes_total, inodes_used@8 as inodes_used, mode@9 as mode, path@10 as path, time@11 as time, total@12 as total, used@13 as used, used_percent@14 as used_percent]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |         DeduplicateExec: [cluster_id@1 ASC,host@5 ASC,time@11 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |           SortPreservingMergeExec: [cluster_id@1 ASC,host@5 ASC,time@11 ASC,__chunk_order@0 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |             SortExec: expr=[cluster_id@1 ASC,host@5 ASC,time@11 ASC,__chunk_order@0 ASC]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |               CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |                 FilterExec: time@11 >= 1687885391398358371 AND (time@11 < -9223372036854775808 OR time@11 > 1656349451399264154)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |                   ParquetExec: file_groups={8 groups: [[5/226/34127/d3be0fff-5e60-471b-aace-0a590adb1a98.parquet, 5/226/34127/575aaf01-a8df-4bb5-b87b-7cff28a67bf1.parquet, 5/226/34127/e1bbf070-088c-4489-ad09-e538da8572fd.parquet, 5/226/34127/eaecefc7-920e-4376-a35d-8a4269756c6c.parquet, 5/226/34127/baf709b4-ec81-4f8a-8017-39612ddbbd2a.parquet, ...], [5/226/34127/0adf9e60-7266-49de-a108-676bab75354f.parquet, 5/226/34127/edba9288-f619-44f9-b1be-0dbe8262a39d.parquet, 5/226/34127/b9d02c61-857e-4009-93c2-d8a1c822dcd1.parquet, 5/226/34127/8066a4c9-6d4a-426a-bcea-09b967ceda49.parquet, 5/226/34127/943f9f8e-08e4-4569-800a-296914c5016b.parquet, ...], [5/226/34127/8cf57581-6b5b-4fab-97a7-1d536d24ed13.parquet, 5/226/34127/5b0e6ea1-823a-450a-9d6f-43d6fc9f8e1e.parquet, 5/226/34127/e7d39be6-b84a-473d-974e-6af1cc177f5c.parquet, 5/226/34127/ee5d0b6a-bd23-4b8b-b8e8-ce3abf81ebb1.parquet, 5/226/34127/3215eb10-4ebf-4ef7-a073-570cd7d1d5a2.parquet, ...], [5/226/34127/42dfc396-74ee-47a7-9083-686afa3e7859.parquet, 5/226/34127/b62c7827-9892-4f97-a20e-debc6b272ba2.parquet, 5/226/34127/65ec07f7-3d83-4cab-8859-cc10968ad192.parquet, 5/226/34127/34f2f8ed-0f44-4b6d-8fc6-fde85ec26b51.parquet, 5/226/34127/9d552b1b-090c-4192-ae04-1ff1963b935c.parquet, ...], [5/226/34127/be2a37fb-c2b4-48f7-959f-b26439e58c6f.parquet, 5/226/34127/a506da75-a2cd-4d74-bdde-5e7737b5ed37.parquet, 5/226/34127/a7035181-b5f3-40b1-be68-c255e7ad77bd.parquet, 5/226/34127/139930db-5e9c-4300-a18f-b9f6eb2a06f9.parquet, 5/226/34127/eb6e703e-e6d3-420d-9c61-23a9c523c84d.parquet, ...], ...]}, projection=[__chunk_order, cluster_id, device, free, fstype, host, inodes_free, inodes_total, inodes_used, mode, path, time, total, used, used_percent], predicate=time@10 >= 1687885391398358371 AND (time@10 < -9223372036854775808 OR time@10 > 1656349451399264154), pruning_predicate=time_max@0 >= 1687885391398358371 AND (time_min@1 < -9223372036854775808 OR time_max@0 > 1656349451399264154) |
```

```SQL
----------
| plan_type    | plan    |
----------
| logical_plan    | Sort: time DESC NULLS FIRST    |
|    |   Projection: h2o.state, h2o.city, h2o.min_temp AS time, h2o.area    |
|    |     TableScan: h2o projection=[area, city, min_temp, state], full_filters=[h2o.min_temp IS NOT NULL]    |
| physical_plan    | SortPreservingMergeExec: [time@2 DESC]    |
|    |   SortExec: expr=[time@2 DESC]    |
|    |     ProjectionExec: expr=[state@3 as state, city@1 as city, min_temp@2 as time, area@0 as area]    |
|    |       UnionExec    |
|    |         CoalesceBatchesExec: target_batch_size=8192    |
|    |           FilterExec: min_temp@2 IS NOT NULL    |
|    |             ParquetExec: file_groups={1 group: [[1/1/1/00000000-0000-0000-0000-000000000000.parquet]]}, projection=[area, city, min_temp, state], output_ordering=[state@3 ASC, city@1 ASC], predicate=min_temp@3 IS NOT NULL    |
|    |         CoalesceBatchesExec: target_batch_size=8192    |
|    |           FilterExec: min_temp@2 IS NOT NULL    |
|    |             RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1    |
|    |               ProjectionExec: expr=[area@1 as area, city@3 as city, min_temp@4 as min_temp, state@5 as state]    |
|    |                 DeduplicateExec: [state@5 ASC,city@3 ASC,time@2 ASC]    |
|    |                   SortPreservingMergeExec: [state@5 ASC,city@3 ASC,time@2 ASC,__chunk_order@0 ASC]    |
|    |                     ParquetExec: file_groups={2 groups: [[1/1/1/00000000-0000-0000-0000-000000000001.parquet], [1/1/1/00000000-0000-0000-0000-000000000002.parquet]]}, projection=[__chunk_order, area, time, city, min_temp, state], output_ordering=[state@5 ASC, city@3 ASC, time@2 ASC, __chunk_order@0 ASC]    |
|    |         CoalesceBatchesExec: target_batch_size=8192    |
|    |           FilterExec: min_temp@2 IS NOT NULL    |
|    |             RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1    |
|    |               ProjectionExec: expr=[area@1 as area, city@3 as city, min_temp@4 as min_temp, state@5 as state]    |
|    |                 DeduplicateExec: [city@3 ASC,state@5 ASC,time@2 ASC]    |
|    |                   SortExec: expr=[city@3 ASC,state@5 ASC,time@2 ASC,__chunk_order@0 ASC]    |
|    |                     RecordBatchesExec: chunks=1, projection=[__chunk_order, area, time, city, min_temp, state]    |
|    |    |
----------
```

Reply:
Here are explains from other tickets

https://github.com/influxdata/EAR/issues/4651

https://github.com/influxdata/EAR/issues/4649

Reply:
> 
> ```sql
> ProjectionExec: expr=[cluster_id@1 as cluster_id, device@2 as device, free@3 as free, fstype@4 as fstype, host@5 as host, inodes_free@6 as inodes_free, inodes_total@7 as inodes_total, inodes_used@8 as inodes_used, mode@9 as mode, path@10 as path, time@11 as time, total@12 as total, used@13 as used, used_percent@14 as used_percent]                      
> ....                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
> |               |       ProjectionExec: expr=[cluster_id@1 as cluster_id, device@2 as device, free@3 as free, fstype@4 as fstype, host@5 as host, inodes_free@6 as inodes_free, inodes_total@7 as inodes_total, inodes_used@8 as inodes_used, mode@9 as mode, path@10 as path, time@11 as time, total@12 as total, used@13 as used, used_percent@14 as used_percent]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
> ```

@NGA-TRAN The first example above contains 2 `ProjectionExec` processes (with unique file_groups), but no `UnionExec` processes. That does indicate they're processed sequentially?

Reply:
@jstirnaman 
I just skimmed the plans and they are too complicated but not complete (becasue we extracted them from customer cases). We will confuse readers.
Give me sometime to finish some other things and then I will create a simple but representative table with data for us to use for explain examples.

If cannot wait a couple of weeks for this, maybe say explain is coming soon or even ignore them in the doc for now?

Reply:
@NGA-TRAN Thanks. I started something similar, see https://github.com/influxdata/docs-v2/pull/5215/files#diff-dd5644e83cdc8dd15e69f44bc117f574a56c4208a688aea384f1d0ebe9a5c0c3R178-R231

Reply:
The lines in the explain was not aligned correctly and leads to different meanings. I will review this more carefully next week (I am overwhelmed right now) but I still think we need a query plan for a clear query example and data if we want customers understand them

Reply:
Agreed. No rush. If you find time and want to work on it together, let me know.

> The lines in the explain was not aligned correctly and leads to different meanings.

That's also the reason that I chose to not use a table-- it mostly fought against preserving whitespace and making it readable.